agent/consul/state/catalog_schema.go

```diff
@@ -284,33 +246,28 @@ func connectNameFromServiceNode(sn *structs.ServiceNode) (string, bool) {
 	}
 }
 
-func indexKindFromServiceNode(raw interface{}) ([]byte, error) {
-	n, ok := raw.(*structs.ServiceNode)
-	if !ok {
-		return nil, fmt.Errorf("unexpected type %T for structs.ServiceNode index", raw)
-	}
-
+func indexKindFromServiceNode(n *structs.ServiceNode) ([]byte, error) {
 	var b indexBuilder
 	b.String(strings.ToLower(string(n.ServiceKind)))
 	return b.Bytes(), nil
 }
 
 // indexWithPeerName adds peer name to the index.
-func indexWithPeerName(
-	fn func(interface{}) ([]byte, error),
-) func(interface{}) ([]byte, error) {
-	return func(raw interface{}) ([]byte, error) {
+func indexWithPeerName[T any](
+	fn func(T) ([]byte, error),
+) func(T) ([]byte, error) {
+	return func(raw T) ([]byte, error) {
+		pi, ok := any(raw).(peerIndexable)
+		if !ok {
+			return nil, fmt.Errorf("wrapped generic type must be peerIndexable: %T", raw)
+		}
+
```

Comment:
Could we remove this type-assertion by constraining `T` to the `peerIndexable` interface like so?

```suggestion
func indexWithPeerName[T peerIndexable](
	fn func(T) ([]byte, error),
) func(T) ([]byte, error) {
	return func(pi T) ([]byte, error) {
```

Reply:
Yep! Matt made a similar comment on the enterprise side. The latest commit should address this